### DISCUSSION TITLE
[Complete] RFC: Incremental Hydration

### DISCUSSION THREAD

**Discussion by thePunderWoman** [angular/angular#57664] [RFCs]
__Authors:__ @thePunderWoman 
__Area:__ Angular Framework
__Posted:__ September 4, 2024
__Status:__ Open

In Angular version 16, we launched full application hydration in Angular to improve your Core Web Vitals, server side rendering performance, and improve the overall initial render experience when using server side rendering. Then, in version 17, we added deferrable views, which allowed you to easily defer less important code until it was needed later, which further improved core web vitals and reduced initial bundle size. We've since also added event replay, which means none of your users' actions will be lost before hydration kicks in.

Building on this functionality, we're excited to share more concrete plans for the next SSR improvement: incremental hydration. With incremental hydration, deferred content is rendered on the server side and skipped over during client-side hydration and bootstrapping - it's left _dehydrated_. This allows Angular to bootstrap the initial page in less time, and keeps the code for those dehydrated components out of your initial bundle. Angular will leave those parts of the UI dehydrated until your declared hydration conditions are met, at which point it will download the necessary code and hydrate those components on demand. You already know the set of hydration triggers, as they're the same familiar `@defer` triggers: `on viewport`, `on interaction`, etc.

Incremental hydration allows you to apply these familiar client-side `@defer` optimizations to server-side rendered code, delaying the loading of less critical parts of your UI while avoiding negative effects like content layout shifts.

## What kind of feedback are we looking for?

While we're excited to hear any and all comments on incremental hydration in Angular, we have additionally called out specific discussion points throughout the RFC for directed feedback. When you leave a comment, be sure to note the open question(s) to which you're responding.

As always, keep our code of conduct in mind. We know that a proposal of this scope significantly impacts Angular and the way you use it in your projects. We appreciate your commitment and investment in this project and ask to keep comments respectful.

## Goals

* Provide a predictable and ergonomic API for developers to leave content dehydrated.
* Reduce initial page load time due to smaller initial bundle sizes.
* Improve core web vitals performance, especially cumulative layout shift.
* Ensure no user interaction is lost prior to fetching and hydration of content.

`Discussion Question 1`: Is it clear to you what is meant by "incremental hydration"?

## Dehydrated Content

When Angular leaves content on the page in a dehydrated state, it behaves as static HTML:

* No component or directive instances are created.
* No lifecycle hooks for the content execute.
* The content does not match view or content children queries.
* The content is not change detected, even if bindings to those components or their contents would have changed. 

In other words, the content appears exactly as it did during server-side rendering.

Only when your hydration condition is met does Angular download the code for the components/directives within the content and hydrate it: instances are created, lifecycle hooks execute, and change detection updates the content based on its bindings.

It's possible that users could interact with your dehydrated content before it can hydrate, especially since the hydration process requires fetching code and is therefore asynchronous. Angular addresses this using the Event Dispatch library, by adding event listeners to your dehydrated content and replaying events after hydration occurs (similar to the `withEventReplay()` option for the main SSR'd application).

## Syntax

Deferrable views, or defer blocks, are the building (defer) block and hydration boundary for incremental hydration in Angular. A defer block becomes a incremental hydration boundary by adding a hydrate on, hydrate when, or hydrate never condition. For example:

```html
<example-cmp />
@defer (on immediate; hydrate on interaction) {
  <deferred-cmp />
  <another-deferred-cmp />
}
```

Similar to defer triggers and prefetch triggers, you can use multiple hydrate triggers at once, and hydration will occur when any of those triggers fire.

Hydration triggers apply when a `@defer` block is server-rendered, during the application's __initial load__. The standard defer block triggers apply only to client side rendered `@defer` blocks. Since any `@defer` block might also be client-side rendered, you will need to configure a regular defer condition in addition to any hydration conditions.

`Discussion Question 2`: Does this syntax make sense to you? Specifically consider the difference between regular triggers and hydrate triggers. Is there an alternative that might work better?

`Discussion Question 3`: Is the defer syntax becoming too hard to read or understand with multiple triggering paradigms and conditions?

`Discussion Question 4`: Do you see yourself commonly using the same triggers for defer and hydrate? Or different conditions? For example: `@defer(on interaction; hydrate on interaction)` vs `@defer(on interaction; hydrate on immediate)`

## Trigger Types

The trigger types for incremental hydration are similar to the main @defer, with a few differences. Hydration triggers always apply to the actual rendered body of the @defer block and thus don't allow element references. They only apply to @defer blocks that are server-side rendered.

### idle

idle triggers hydration once the browser has reached an idle state (see `requestIdleCallback`).

### interaction
`hydrate on interaction` instructs Angular to hydrate content when the user interacts with it (click, focus, touch, and input events like keydown, blur, etc).

### immediate
`hydrate on immediate` triggers the deferred load and hydration immediately.

### timer(x)
timer(x) triggers hydration after a specified duration.

### hover
hover triggers deferred loading when the user hovers their mouse over the server rendered content. This condition is useful if hovering the mouse is a good indication that the user might soon interact with the content.

### viewport
viewport triggers hydration when the server rendered content enters the viewport (detected using the `IntersectionObserver` API).

### never
`hydrate never` tells Angular to leave your component dehydrated indefinitely, effectively treating it as static content. Note that this is not the same as a server-only component. When a `@defer` block is instantiated during normal client-side rendering, the content is still fetched and rendered normally.

### when
Just like regular and prefetch deferrable view triggers, you can provide a custom condition which triggers hydration when it becomes `true`. There are some caveats to this, which is outlined in the nesting section.

## Nested Hydration

When multiple `@defer` blocks are in a dehydrated state, their conditions are applied simultaneously. For example, in the following structure:

```js
@defer (hydrate on hover) {
  @defer (hydrate on timer(15s)) {
    ...
  }
}
```

The outer block will hydrate if the mouse is hovered over any of the content inside of it. Even if the block is never hovered, the inner block's timer will trigger hydration after 15 seconds.

There is one exception to this behavior: `hydrate when` only works if the component which contains it is already hydrated. If a `when` trigger for a block is nested inside a component which is itself dehydrated, it won't be evaluated until the parent component is triggered and hydrated independently. For example:

```js
@defer (hydrate on hover) {
  @defer (hydrate when user() !== null) {
    ...
  }
}
```

If the outer block has not triggered, then regardless of the value of `user()`, the inner block will not hydrate.

When hydration is triggered within a nested block, any parent block(s) are first hydrated if necessary, and then the triggered block is hydrated. Because each hydration step requires fetching the code for dependencies, this means that nested dehydrated blocks require a waterfall of code loading in order to hydrate.

`Discussion Question 5`: Do you have components you would like to use incremental hydration with but would be unable to due to the potential for stale data? Would that change if incremental hydration was triggered by inputs changing? Similarly would that change if hydration was triggered by signals changing?

### Root text nodes

The hydration triggers of `interaction`, `hover`, and `viewport` do not work with plain text nodes at the top (root) level of the deferred content, as browsers do not support listening for events on text nodes or watching them with `IntersectionObserver`. Angular will therefore not allow such nodes in `@defer` blocks which use those triggers. Text must be wrapped in an element (such as `<p>`) in order to work correctly with incremental hydration.

# Frequently Asked Questions

## Q: I have a question, but it's not ready yet.
…ok

## Q: When will this feature be available?
When it's fully hydrated.

## Q: Islands?
We're thinking about them.

## Q: Does this mean Angular has server components now?
No, it does not. Server components should always render on the server and never on the client. hydrate triggers only apply to first load.

## Q: When hydration is triggered on nested blocks, each with hydrate triggers, will that result in cascading requests or waterfalls?
Yes, hydrate triggers on nested components will be a waterfall of requests, but this is something we plan to address.

## Q: OK, my question is ready now. Does this new version of Angular Cereal also have marshmallows?
Ah I get it, this is a hydration trigger joke…and yes…marshmallows.

**Comment by t1m4lc** [comment]
Answer Question 1: No 🫠 please explain and give us some use case

**Reply by thePunderWoman** [reply]
Use case A:
You have a list of product items on a list page on an ecommerce platform. You don't know if a user will ever see or interact with that product. So you put it underneath a defer block with hydrate syntax for it to fetch and hydrate only once it's been interacted with, reducing your initial bundle size. In this case, the product still gets rendered, but the code for the component doesn't get fetched until it's needed.

Use case B:
You have a blog post that is mostly static content. You can use a `hydrate never` condition to avoid shipping any javascript for that post component.

Use case C:
You have a large component above the fold, like your header or carousel. Your heatmaps show that very few people click on those. So you can put them behind a hydrate trigger to render them, but avoid shipping the content until someone interacts with it.

**Reply by hheexx** [reply]
@thePunderWoman maybe deferred hydration is better name? It implies it's related to defer blocks......

**Reply by mgechev** [reply]
We discussed deferred hydration as a potential name. It's possible to perform full-app hydration on idle, which makes it deferred, but not necessarily incremental. Calling it "incremental deferred hydration" will be accurate, but also too long without bringing too much value.

**Reply by naveedahmed1** [reply]
Incremental Hydration and Partial Hydration both seems natural and reflects exactly what the feature is about.

**Comment by ilirbeqirii** [comment]
Question (2): What would be the execution flow of the provided example with a mix of standard and hydration triggers? Syntax is not hard to read but how to reason about it (execution)

```
@defer (on immediate; hydrate on interaction) {
  <deferred-cmp />
  <another-deferred-cmp />
}
```

**Reply by thePunderWoman** [reply]
Think of the regular defer trigger, in this case `on immediate` as your client side rendering trigger only. The `hydrate on interaction` applies to when this defer block appears in a server render. So think about it in terms of your initial load trigger.

So if it was your initial server side render, the interaction trigger applies. If a user was on a different page and navigated via a client side render to the page with this code, the immediate trigger would apply.

**Reply by ilirbeqirii** [reply]
If this means that, for the **initial server-side render**, the content on the **defer block** will be a **static HTML** that will be hydrated on user interaction, and the regular/client-side rendering trigger is not taken into account then we can say that trigger types (regular & hydration) are **mutually exclusive**?

Thanks in advance for all the great work you are doing 🅰❤

**Reply by thePunderWoman** [reply]
**mutually exclusive** is a good way to put it, yes. Hopefully that's clear and makes sense. We are also utilizing event replay for the content in the hydrate blocks. So if a user interacts with that content, the interaction will not be lost.

**Reply by zygarios** [reply]
So I understand that the 'Hydrate' syntax makes the 'dom' structure render without hydration. But what happens when I use @placeholder or @placeholder+@loader together with this function? Does that mean they will be skipped on the server side?

**Reply by thePunderWoman** [reply]
> So I understand that the 'Hydrate' syntax makes the 'dom' structure render without hydration. But what happens when I use @Placeholder or @Placeholder+@loader together with this function? Does that mean they will be skipped on the server side?

This is correct. In the SSR / hydrate case, your defer main template content **IS** your new placeholder. The `@placeholder` and `@loading` templates you provide will be used in the client side rendering cases, just like they do today.

**Reply by zygarios** [reply]
Ok, but if I don't use the word 'Hydrate', the placeholder will come from the server. I just want to be sure.

**Reply by thePunderWoman** [reply]
@zygarios If you don't use the word `hydrate`, things will behave exactly as they do now with SSR. The `@placeholder` block gets rendered on the server and is hydrated eagerly with the rest of the full application.

**Comment by e-oz** [comment]
I'm super happy to see this RFC, everything here makes sense. I have just one feature request: please consider adding the “time” argument to “idle”.
Like:
```
@defer (on hover(container); on idle(2000)) {
  <heavy-component/>
}
```

That would load the HeavyComponent either on hover or after a 2s delay after “idle”.

There are just components that are far enough from the user’s path, and heavy enough to load them not on just “idle”, but with a small delay - because more important components will take a chance to be loaded on “idle”.

Overall, this RFC is awesome and I hope it will be implemented in v19.

**Reply by crisz** [reply]
I'd advise `idleTimeout` or something else that makes clear it's a timeout. `idle(2000)`, with no context, may not convey enough information about the behaviour

**Reply by thePunderWoman** [reply]
@e-oz This is definitely more of a feature request for `@defer` as a whole, rather than anything specific to this hydration RFC. Could you create a issue / feature request with this instead?

**Reply by e-oz** [reply]
> Could you create a issue / feature request with this instead?

Sure! https://github.com/angular/angular/issues/57814

**Comment by YonatanZviLevy** [comment]
Will there be support for placeholder and loading blocks?

**Reply by thePunderWoman** [reply]
This is additional syntax for `@defer` blocks. So `@placeholder`, `@loading`, and `@error` are all supported. However they wouldn't apply for hydrate cases as the placeholder is the server rendered content, and there's no need to display loading, as the actual content is already present. They'd still work as they normally would in a CSR situation though.

**Comment by zygarios** [comment]
Discussion Question 1: Yes, it's clear
Discussion Question 2: After reading the comments, it is clear to me. However, I think it would be helpful if there were an **infographic** or well-described scenarios explaining what happens in different contexts. An example is how @placeholder and @loader behave in combination with hydrate. I'm not just looking for an explanation of hydrate, but rather the entire flow and use cases for the defer functionality as a whole.

Discussion Question 3: The syntax doesn't seem to be overly complicated. However, it is important to clearly describe the dependencies between different states, as I mentioned in point 2.

Discussion Question 4: I don't have an opinion on this matter

Discussion Question 5: I don't have an opinion on this matter

**Comment by samuelfernandez** [comment]
The RFC is great, good job!

* Question 1: Yes
* Question 2: Yes
* Question 3: A little verbose, bug gives full control. Would it make sense to provide defaults via config?
* Question 4: It's possible. But this whole feature is about providing tooling to fit a variety of conditions, web apps are complex. The proposal can enable any use case and leave the developer optimize depending on use case.
* Question 5: When using defer, it might not be obvious that, if the view depends on some external state, dehydrated views might be stale. Ideally the framework would solve that and force hydration if it detects that the view is stale because of a signal update, just as qwik does. If not, I feel some diagnostic errors to avoid the pitfall would be needed.

Additional feedback:

* Hydration needs to fetch chunks of code, which come with a cost. Loading and parsing can impact user experience. Ideally a solution similar to qwik's could be provided: a Service Worker in the background preloads chunks in a separate thread, and when needed they are available for the main thread to hydrate without the network call or parsing time.
* One of the biggest wins from this new feature is leaving content dehydrated. This can be much beneficial, since there are parts of a website that are just static. For example, sidenav, header, footer. Usually those parts of the app contain only links. However, a dehydrated link that used `routerLink` would not work in the SPA because it does not integrate well with the router. In order to enable apps have more dehydrated content, I'd like to recommend researching other solutions for router links. One realistic way would be that metadata for the link is left in the DOM, use a regular `href`, and through event bubbling capture the event at a root component and handle navigation within Angular's router. This way we'd benefit from staying within the SPA routing, yet leaving links and their components dehydrated.

**Reply by thePunderWoman** [reply]
@samuelfernandez RE: RouterLink - Yeah, this is something we've thought about a bit, but haven't fully explored. We agree that it would be a nice to have. 

RE: service worker - we've considered this too. This is something that could be a future enhancement, among other things we're considering.

RE: Question 3: We've also thought about the config option, but the concern was that a config would make it harder to look at a defer block and reason about what it would do without knowing about that config. If we get strong signal that this is desired, it's certainly something we could add.

RE: Question 5: Signal updates and dehydrated content are something we're definitely thinking a lot about. Expect to hear more about this in the future.

**Reply by osnoser1** [reply]
When I read about this RFC, the first use case I imagined could be a potential win was routerLink handling in dehydratated content; I'd love this

**Reply by thePunderWoman** [reply]
As a quick follow up, the only real concern around routerLinks in dehydrated code would be when routerLinks exist in `hydrate never` blocks. Any interaction with a routerLink in any other case will trigger hydration, and that click will get replayed, resulting in the expected client side navigation. In a `hydrate never` case, as long as routerLink is used on an anchor tag, it would navigate, but as a hard navigation. Any other use of routerLink, though, wouldn't do anything inside a `hydrate never` block.

**Reply by samuelfernandez** [reply]
Exactly, that is why it'd be wonderful that in hydrate never it'd still trigger an SPA navigation without needing to hydrate. For simple use cases shouldn't be that complex. The event of a hard navigation could be intercepted in the root element, the url parsed, and router execution triggered. It's nice to see that it may be explored 😃. This way only m truly interactive regions would need to be hydrated,

**Reply by samuelfernandez** [reply]
@thePunderWoman One additional question on this statement:

> Any interaction with a routerLink in any other case will trigger hydration, and that click will get replayed, resulting in the expected client side navigation.

`routerLink` implementation does `event.preventDefault()`, which cancels browser's hard navigation, and does a SPA navigation. When the component is not yet hydrated, what does exactly happen? Does the event replay library also call `event.preventDefault()`? If not, the browser will synchronously execute a hard navigation, and hydration will just be lost. Asking without actually testing it... 😅 Thanks for helping me understand the deep waters of hydration 💧 (pun intended).

**Reply by thePunderWoman** [reply]
@samuelfernandez The event replay behavior makes sure the original click is not a hard navigation in this case. So it will behave as you would want, with the replay and the client side navigation.

**Comment by e-oz** [comment]
> Discussion Question 1: Is it clear to you what is meant by "incremental hydration"?

A little bit mixed with `prefetch`, but it's ok.

> Discussion Question 2: Does this syntax make sense to you? Specifically consider the difference between regular triggers and hydrate triggers. Is there an alternative that might work better?

Syntax makes complete sense.

> Discussion Question 3: Is the defer syntax becoming too hard to read or understand with multiple triggering paradigms and conditions?

No. Especially if Angular warns us about non-usable combinations, as it does with `@defer (on viewport)` without an element ref and placeholder.

> Discussion Question 4: Do you see yourself commonly using the same triggers for defer and hydrate? Or different conditions? For example: @defer(on interaction; hydrate on interaction) vs @defer(on interaction; hydrate on immediate)

No, but I do see common use with `prefetch`.

> Discussion Question 5: Do you have components you would like to use incremental hydration with but would be unable to due to the potential for stale data? Would that change if incremental hydration was triggered by inputs changing? Similarly would that change if hydration was triggered by signals changing?

I don't think so.

**Reply by thePunderWoman** [reply]
@e-oz So `prefetch` just makes sure your chunk is prefetched, but your defer block still renders what's in the @placeholder block until triggered. `hydrate` is a signal to the server side that you'd like to render your defer block main template on the server. `hydrate` changes what is rendered on the server server side, affecting what the user sees on first render. `prefetch` specifies how early the chunk is fetched on the client side. Does that help clarify?

**Reply by e-oz** [reply]
Sure. I meant that `prefetch` is designed to run before the actual execution, just like `hydrate`. But I do realize that `hydrate` does much more than `prefetch`. My point was - maybe `prefetch` already points to things we would like to `hydrate`. But it's merely an assumption; I wasn’t trying to say anything against the idea itself. The idea is awesome, and I hope to try it as soon as possible - before February 30, 2025. 😎

**Comment by naveedahmed1** [comment]
1. Regarding `Nested Hydration`, is there any specific reason we need two different behaviors for `hydrate when` and other `hydrate triggers`?

2. What will happen in this case?
   
```
@defer (hydrate on hover) {
...
..  
@defer (hydrate never) {
    <ad-unit>
  }
...
...
}
```
I'd want the outer component to be hydrated while keeping the ad-unit component unhydrated.

3. Regarding the `Syntax`:

What if some of the hydration triggers could also be added to Component config, so that whenever we use the component it's default hydration settings is applied, with option to overwrite with `@defer` block.

```
@Component({
  templateUrl: './enable-notifications.component.html',
  standalone: true,
hydrate: [
interaction
],
prefetch:true
})
```

**Reply by thePunderWoman** [reply]
1. Yes there is. The reason being that the rest of the triggers are _event_ based and `hydrate when` is _expression_ based. _Event_ based triggers do not rely on logic in the component to trigger and can have shared logic for all of the instances across your page. _Expressions_ rely on logic in your component that contains the defer block. We cannot evaluate an expression that does not exist yet. If the immediate parent is hydrated, the _expression_ will be present and can be evaluated. If the parent is dehydrated, that logic hasn't yet been fetched.

2. Your expectation is correct and how things should behave on the initial SSR page load. Subsequent page loads would result in that logic being fetched, as subsequent page loads are client side rendered.

3. We did consider components as hydration boundaries and also defer boundaries, but since the entire component is not what is being deferred, this sort of syntax wouldn't work, as you can have multiple defer blocks in a given component template. We're also actively trying to reduce the amount of configuration in the component decorator. So this would be moving us in the opposite direction of that goal.

**Comment by brandonroberts** [comment]
Q2: Any reason `hydrate` can't be used as its own block, with the deferred behavior being implied?

```ts
@hydrate (on hover) {
    <my-component />
}
```

**Reply by thePunderWoman** [reply]
We've certainly considered this, but what does it do on a Client Side Render in this case? `@defer` blocks render in CSR and SSR. I would assume you'd still want the content to appear during CSR when a `@hydrate` block is present.

**Reply by brandonroberts** [reply]
Yes, I would assume Client Side Render could still use a `@placeholder`.

**Reply by thePunderWoman** [reply]
@brandonroberts but since it's called @hydrate, which is specifically a server side rendered behavior, does that not send a confusing message? You can't hydrate client side rendered content.

**Reply by brandonroberts** [reply]
Ok, I misunderstood the connection with CSR then. I was thinking of this in terms of shorthand syntax that works in both scenarios.

I thought if you happen to use `@defer` and hydrate without server-side rendering the page first, it would fall back to just being deferrable. Will it throw an error or warning if so?

**Reply by thePunderWoman** [reply]
It wouldn't make sense to use defer and hydrate conditions without SSR. Similarly it wouldn't make sense to use `@hydrate` in that scenario. What I'm getting at is these hydrate conditions apply to SSRed content, which means first load. It may be unclear to folks that they're affecting CSR situations when using `@hydrate` when by its nature `hydration` is an SSR thing. `@defer` with hydrate conditions may be clearer to people that it serves both cases.

**Reply by naveedahmed1** [reply]
I agree that @defer is more appropriate. In case of partial hydration we are in fact deferring the hydration to a more appropriate time.

**Comment by yjaaidi** [comment]
> Discussion Question 4: Do you see yourself commonly using the same triggers for defer and hydrate? Or different conditions? For example: @defer(on interaction; hydrate on interaction) vs @defer(on interaction; hydrate on immediate)

I guess that I would generally use different triggers as in most cases, placeholders are short-lived as they don't make as much sense to the user as a dehydrated view which can often be hydrated way later.

**Reply by naveedahmed1** [reply]
Could you please provide an example where you have a server-side rendered (SSR) page, and you'd like to hydrate a child component based on a specific trigger? Additionally, in the case where the user navigates to the same page via client-side navigation, you want that component to be defer-loaded using a different trigger. I almost always ended up using the same trigger for both cases.

**Reply by yjaaidi** [reply]
A heavy datepicker that users don't often interact with.
If SSRed, I'd want it to be hydrated on interaction as it is unnecessary to hydrate it if it's not used.
On the other hand, if it's CSRed, I would want to replace my placeholder on idle because the date it displays is relatively useful to the user even though they do not interact with it.

**Reply by naveedahmed1** [reply]
That’s an interesting use case, and while you might have specific requirements, I would generally avoid server-side rendering a datepicker. When a datepicker is rendered on the server, it might display the server's date and time, which could be confusing or inaccurate for users in different time zones. This could lead to errors if the displayed date doesn't match the user's local time. So, if you do SSR a datepicker, it’s crucial to hydrate it as soon as possible or during idle time to ensure the correct date is shown. However, I understand you may have unique needs.

**Reply by godhelpme1993** [reply]
I think about it alot. walking around not trusting anyone

**Comment by yjaaidi** [comment]
My gut feeling is that most users would prefer the default behavior to be `hydrate on idle` instead of SSRing the placeholder.

Is there a reason why `hydrate on idle` is not the default behavior set apart that it would be a breaking change?
This would require some syntax to skip SSR (`skip-ssr`?) and a migration to enable it on all existing `@defer`s so that users can opt-in.

**Reply by naveedahmed1** [reply]
I support this idea. The below code should hydrate on idle by default.

```
@defer {
  <my-component />
}
```

> SSRing the placeholder

Placeholders are not server-side rendered (SSRed). With partial hydration, the main template of the deferred block is rendered on the server and serves as a placeholder until the component is hydrated. During client-side navigation, your placeholder block would be used instead (the same way it currently works).

**Reply by thePunderWoman** [reply]
Yes there is a reason we didn't make `hydrate on idle` the default. People may not want some blocks to use partial hydration, and if we have it default when omitted, there's no way to avoid it. We have considered a shorthand, though, which would be something like `@defer(on idle; hydrate)`. Where essentially just saying `hydrate` with no trigger indicates you want to use the same trigger as the regular defer. You could omit `on idle` in that case and it would just be `@defer(hydrate)`. How do you feel about that concept?

**Reply by naveedahmed1** [reply]
@defer block doing hydrate on idle by default feels more natural to me 😀

**Reply by yjaaidi** [reply]
💯 for `@defer(hydrate)` shorthand.

> People may not want some blocks to use partial hydration, and if we have it default when omitted, there's no way to avoid it.

There could be an instruction to opt-out from hydration, even though it will be tough to name 😅: `hydrate disabled`? it's going to be hard to find something that doesn't get confusing with `hydrate never`

**Reply by alxhub** [reply]
Perhaps something like:
```
@defer (on idle; server placeholder) {...}
```
it doesn't need to be a `hydrate` condition specifically...

**Reply by naveedahmed1** [reply]
I am not sure what `server placeholder` would do in this case?

**Comment by yjaaidi** [comment]
> Discussion Question 3: Is the defer syntax becoming too hard to read or understand with multiple triggering paradigms and conditions?

Yes. I am afraid that this can become quite confusing, especially for newcomers. 
Good default values can help by moving the defer options further in the learning path.

Another thing that could help is something like reusable policies:
```ts
// defer-policies.ts
const lowPriorityPolicy = createDeferPolicy({defer: 'idle', hydrate: 'interaction'});

// my-cmp
@Component({
  template: `
  @defer(policy lowPriorityPolicy) {
    ...
  }
  `
})
class MyCmp {
  lowPriorityPolicy = lowPriorityPolicy
}
```
or a better syntax if possible.

The policies would of course have to be static so that the compiler can understand them.

@thePunderWoman if you think this could be interesting, I can open an issue for this as it is not directly related to the RFC.

**Reply by thePunderWoman** [reply]
Being able to create reusable policies is definitely a thing we've considered. So if there's enough interest, it could be a future feature add.

**Comment by naveedahmed1** [comment]
What if I want the whole app to be set as hydrate never and then hydrate individual child components on demand e.g. hydrate some components on interaction and some on when they enter the view port. Something like an app wide confg that makes the whole app a static app with almost zero JavaScript unless a user starts interacting with the app.

**Reply by thePunderWoman** [reply]
That concept is essentially what resumability is. We love this idea, but it is quite a bit more challenging and complicated to get to this point. Dependency injection, data dependencies, and a number of other complications all factor into making this a workable feature with Angular. We think incremental hydration is a step towards that possibility in the future. So stay tuned.

**Comment by mixi342** [comment]
I use an ngx-image-cropper component, that access window, and document objects right at the start.

In Angular SSR application this component throws an error: "ERROR RuntimeError: NG04002.

Two solutions helped me:

@defer (on immediate) {
   <image-cropper...
}

or

   <image-cropper ngSkipHydration

With incremental hydration, what solution should I choose?

**Reply by thePunderWoman** [reply]
Ideally you would update `ngx-image-cropper` to use `afterRender` hooks to make them safe for hydration. In lieu of that, if you use a hydrate trigger, I'd go with `ngSkipHydration` on this element, as it's no longer client side rendered. Alternatively if this is a component that only ever has the image cropper in it, just skip using a hydrate trigger and leave it like it is, as that's not going to change behavior.

**Comment by naveedahmed1** [comment]


**Reply by thePunderWoman** [reply]
Sorry. RFC conversations don't typically discuss expected timelines.

**Comment by muhammadawaisshaikh** [comment]
Sounds like an amazing game 🤩 Loge to see the future of apps with Angular. This is very amazing chunk to even not load KB's of code until we interact.

**Comment by thePunderWoman** [comment]
tl;dr; thank you for all the feedback - our intention is to implement APIs described in this RFC and roll them out in developer preview for Angular v19!

## Shipping it as a developer preview in Angular v19!
Thank you so much to all of you that left comments, questions, and feedback in this RFC. It was helpful in understanding people's impressions of this API and functionality.

## Summary of the discussion questions:

1. Is it clear to you what is meant by "incremental hydration"?

Mostly the answer here was "yes", but there were a few clarifying questions people asked.

2. Does this syntax make sense to you? Specifically consider the difference between regular triggers and hydrate triggers. Is there an alternative that might work better?

The answer here is also largely "yes it makes sense", with some suggestions for a `@hydrate` block as an alternative. We had a few questions about execution between trigger types, which were clarified in the comments.

3. Is the defer syntax becoming too hard to read or understand with multiple triggering paradigms and conditions?

The answer here is that the syntax is clear, but there might be confusion around the behavior between the two trigger types. We hope to address this with clear documentation.

4. Do you see yourself commonly using the same triggers for defer and hydrate? Or different conditions?

There wasn't much response here to indicate there were concerns.

5. Do you have components you would like to use incremental hydration with but would be unable to due to the potential for stale data? Would that change if incremental hydration was triggered by inputs changing? Similarly would that change if hydration was triggered by signals changing?

Similarly there wasn't much response to this question. I think we'll likely get more feedback as people try this feature out.

## Wrap up
Based on all the comments and feedback we didn't find any use cases that were not covered in this design. There's some interest in a specific `@hydrate` block, but we felt that the hydrate syntax for defer blocks was a better step forward.
We've included this feature in v19 as a developer preview API. So we're looking forward to your usage and feedback of incremental hydration with the upcoming RCs, final release, and beyond.

Thanks everyone!
